core/src/main/kotlin/com/strumenta/kolasu/model/Indexing.kt

```diff
@@ -0,0 +1,49 @@
+package com.strumenta.kolasu.model
+
+import java.util.IdentityHashMap
+
+fun interface IdProvider {
+    fun getId(node: Node): String?
+}
+
+class SequentialIdProvider(private var counter: Int = 0) : IdProvider {
+    override fun getId(node: Node): String? {
+        return "${this.counter++}"
+    }
+}
+
+class OnlyReferencedIdProvider(
+    private var idProvider: IdProvider = SequentialIdProvider()
+) : IdProvider {
+    override fun getId(node: Node): String? {
+        val root = (node.walkAncestors().lastOrNull() ?: node).apply { assignParents() }
+        return root.walk().find { otherNode ->
```

Comment:
We traverse the whole AST at each request. I think we should either pre-calculating which nodes are referred to at construction time, or cache results

Reply:
This should be covered now